text/0000-implied-bounds.md

```diff
@@ -0,0 +1,739 @@
+    let a = <T as Foo>::Item;
+}
+```
+
+### **Inherent impls**
+Given an inherent impl:
+```rust
+impl<r..., X1, ..., Xn> SelfTy where WhereClause1, ..., WhereClausek {
+    /* body of the impl inside here */
+}
+```
+We rely on the following assumptions inside the body of the impl:
+* `Expanded({ WhereClause1, ..., WhereClausek })`
+* `WF(T)` for all `T ∈ InputTypes(SelfTy)`
+* `WF(Xi)` for all `i`
+
+Methods make their normal assumptions + the set of assumptions made by the impl. Things to prove inside methods do not differ from normal fns.
+
+A caller of a method has to prove that the where clauses defined on the impl hold, in addition to the requirements for calling general fns.
+
+## Proving well-formedness for input types
+[proving-wf-input-types]: #proving-well-formedness-for-input-types
+
+One would have noticed that we only prove well-formedness for input types in a lazy way (e.g., inside function bodies). This means that if we have a function:
+```rust
+struct Set<K: Hash> { ... }
+struct NotHash;
+
+fn foo(arg: Set<NotHash>) { ... }
+```
+then no error will be caught until someone actually tries to call `foo`. Same thing for an impl:
+```rust
+impl Set<NotHash> { ... }
+```
+the error will not be caught until someone actually uses `Set<NotHash>`.
+
+The idea is, when encountering an fn/trait impl/inherent impl, retrieve all input types that appear in the signature / header and for each input type `T`, do the following: retrieve type variables `X1, ..., Xn` bound by the declaration and ask for `∃X1, ..., ∃Xn; WF(T)` in an empty enviromnent (in Chalk terms). If there is no possible substitution for the existentials, output a warning.
+
+Example:
+```rust
+struct Set<K: Hash> { ... }
+
+// `NotHash` is local to this crate, so we know that there exists no `T`
+// such that `NotHash<T>: Hash`.
+struct NotHash<T> { ... }
+
+// Warning: `foo` cannot be called whatever the value of `T`
+fn foo<T>(arg: Set<NotHash<T>>) { ... }
+```
+
+## Cycle detection
+In Chalk this design often leads to cycles in the proof tree. Example:
+```rust
+trait Foo { }
+// `WF(Self: Foo)` holds.
+ 
+impl Foo for u8 { }
+
+// Expanded to `trait Bar where Self: Foo, WF(Self: Foo)`
+trait Bar where Self: Foo { }
+
+// WF rule:
+// `WF(Self: Bar)` holds if `Self: Foo`, `WF(Self: Foo)`.
+
+// Reverse WF rules:
+// `Self: Foo` holds if `WF(Self: Bar)`
+// `WF(Self: Foo)` holds if `WF(Self: Bar)`
+```
+Now suppose we are asking wether `u8: Foo` holds. The following branch exists in the proof tree:
+`u8: Foo` holds if `WF(u8: Bar)` holds if `u8: Foo` holds.
+
+I *think* rustc would have the right behavior currently: just dismiss this branch since it only leads to the tautological rule `(u8: Foo) if (u8: Foo)`.
+
+In Chalk we have a more sophisticated cycle detection strategy based on tabling, which basically enables us to correctly answer "multiple solutions", instead of "unique solution" if a simple *error-on-cycle* strategy were used. Would rustc need such a thing?
```

Comment:
I think the case tabling comes into play is:
```Rust
trait Bar {} // not implemented for anything
trait Foo {}
impl<T: Foo + #[cfg(maybe)] Bar> Foo for Vec<T> {}
#[cfg(maybe2)]
impl Foo for () {}
```

And then we ask whether `?0: Foo`. Currently, rustc will answer "ambiguous" for all 4 cases, while tabling might be able to figure out a more precise answer.

Reply:
Yes exactly, there is [a test](https://github.com/nikomatsakis/chalk/blob/master/src/solve/test.rs#L284-L304) in Chalk for that precise example.

Reply:
I think that to be "concrete" the example also needs
```Rust
struct u32 {}
impl Bar for u32 {}
```

To avoid the prover just seeing that `Bar` has no possible impls.